PL2/PL2-parser/src/main/java/nl/tudelft/pl2016gr2/parser/controller/FullGfaReader.java

```diff
@@ -3,153 +3,217 @@
-    originalGraph = new OriginalGraph();
-    prepNodes(originalGraph, numNodes);
-    read();
-
-    sc.close();
-    sc = null;
+  public FullGfaReader(String fileName) {
+    read(fileName);
   }
 
   /**
-   * Initialize the nodes of a graph.
+   * Read a GFA file.
    *
-   * @param graph the graph of which to initialize a set amount of node.
-   * @param nodes the amount of nodes to initialize.
+   * @param fileName the name of the GFA file.
    */
-  protected static void prepNodes(OriginalGraph graph, int nodes) {
-    for (int i = 1; i < nodes + 1; i++) {
-      graph.addNode(new Node(i, 1, new ArrayList<>(), 0));
+  private void read(String fileName) {
+    try {
+      BufferedReader br = new BufferedReader(new InputStreamReader(FullGfaReader.class
+          .getClassLoader().getResourceAsStream(fileName)));
+      br.readLine();
+      String line;
+      while ((line = br.readLine()) != null) {
+        char[] chars = line.toCharArray();
+        switch (chars[0]) {
+          case 'L':
+            parseEdge(chars);
+            break;
+          case 'S':
+            parseNode(chars);
+            break;
+          case 'H':
+            parseHeader(chars);
+            break;
+          default:
+        }
+      }
+    } catch (IOException ex) {
+      Logger.getLogger(FullGfaReader.class.getName()).log(Level.SEVERE, null, ex);
     }
   }
-  
+
   /**
-   * Read the actual data.
+   * Parse a line of the GFA file which contains an edge.
+   *
+   * @param chars characters of the line.
    */
-  private void read() {
-    int count = 0;
-    while (sc.hasNext()) {
-      count++;
-      String next = sc.next();
-      switch (next) {
-        case "H":
-          handleH(sc);
-          break;
-        case "S":
-          handleS(sc);
-          break;
-        case "L":
-          handleL(sc);
-          break;
-        default:
-          System.out.println("Not S, H or L at: " + count);
-          sc.nextLine();
-      }
+  private void parseEdge(char[] chars) {
+    int from = 0;
+    int index = 2;
+    while (chars[index] != '\t') {
+      from *= 10;
+      from += chars[index++] - 48;
     }
-    System.out.println("Number of lines read: " + count);
+    index += 3;
+    int to = 0;
+    while (chars[index] != '\t') {
+      to *= 10;
+      to += chars[index++] - 48;
+    }
+    getNode(to).addInlink(from);
+    getNode(from).addOutlink(to);
   }
 
   /**
-   * Private method for handling an "H" as first character on the line of a gfa file.
-   * @param sc The scanner which is used to parse the file.
+   * Parse a line of the GFA file which contains a node.
+   *
+   * @param chars characters of the line.
    */
-  private void handleH(Scanner sc) {
-    String line = sc.nextLine();
-    this.genoms = extractGenomes(line);
+  private void parseNode(char[] chars) {
+    int nodeId = 0;
+    int index = 2;
+    while (chars[index] != '\t') {
+      nodeId *= 10;
+      nodeId += chars[index++] - 48;
+    }
```

Comment:
Same thing about magic numbers here. Also noticing that the while loop seems identical. If the values in the loop (`10` and `48`) are independent of the method where they are used (`parseNode` or `parseEdge`), but based on some external factor, such as data structure (the char array) or the file format, this should definitely be separated from the methods. If these values have to change we will now need to do so in both methods, which is very error prone. (especially because it is such a small block).

A class-wide constant is sufficient, but a separate method is could be even better, because also keeps file length down.